core/vmx_ops.asm

```diff
@@ -0,0 +1,364 @@
+    cmp reg_arg2_16, 1h
+    mov rax, reg_arg1
+    mov rcx, [rax + vcpu_state._rcx]
+    mov rdx, [rax + vcpu_state._rdx]
+    mov rbx, [rax + vcpu_state._rbx]
+    mov rbp, [rax + vcpu_state._rbp]
+    mov rsi, [rax + vcpu_state._rsi]
+    mov rdi, [rax + vcpu_state._rdi]
+    mov r8,  [rax + vcpu_state._r8]
+    mov r9,  [rax + vcpu_state._r9]
+    mov r10, [rax + vcpu_state._r10]
+    mov r11, [rax + vcpu_state._r11]
+    mov r12, [rax + vcpu_state._r12]
+    mov r13, [rax + vcpu_state._r13]
+    mov r14, [rax + vcpu_state._r14]
+    mov r15, [rax + vcpu_state._r15]
+    mov rax, [rax + vcpu_state._rax]
+    je .resume
+    vmlaunch
+    jmp .exit_entry_fail
+.resume:
+    vmresume
+    jmp .exit_entry_fail
+.exit_entry:
+    push rdi
+    mov rdi, [rsp+8]
+    mov [rdi + vcpu_state._rax], rax
+    mov [rdi + vcpu_state._rcx], rcx
+    mov [rdi + vcpu_state._rdx], rdx
+    pop rcx
+    mov [rdi + vcpu_state._rbx], rbx
+    mov [rdi + vcpu_state._rbp], rbp
+    mov [rdi + vcpu_state._rsi], rsi
+    mov [rdi + vcpu_state._rdi], rcx
+    mov [rdi + vcpu_state._r8], r8
+    mov [rdi + vcpu_state._r9], r9
+    mov [rdi + vcpu_state._r10], r10
+    mov [rdi + vcpu_state._r11], r11
+    mov [rdi + vcpu_state._r12], r12
+    mov [rdi + vcpu_state._r13], r13
+    mov [rdi + vcpu_state._r14], r14
+    mov [rdi + vcpu_state._r15], r15
+.exit_entry_fail:
+    ; pop the state
+    pop rbx
+    pop rbx
+    pop rax
+    pop rbp
+    pop rdi
+    pop rsi
+    pop rdx
+    pop rcx
+    pop r15
+    pop r14
+    pop r13
+    pop r12
+    pop r11
+    pop r10
+    pop r9
+    pop r8
+    vmx_check
+    popfq
+    ret
+%elifidn __BITS__, 32
+    pushfd
+    push ecx
+    push edx
+    push esi
+    push edi
+    push ebp
+    push eax
+    push ebx
+    ; write host rsp
+    mov ebx, 6C14h
+    mov eax, esp
+    sub eax, 4h
+    vmwrite ebx, eax
+    pop ebx
+    pop eax
+    push eax
+    push ebx
+    ; push the state
+    mov eax, reg_arg1
+    push eax
+    cmp reg_arg2_16, 1h
+    mov ecx, [eax + vcpu_state._rcx]
+    mov edx, [eax + vcpu_state._rdx]
+    mov ebx, [eax + vcpu_state._rbx]
+    mov ebp, [eax + vcpu_state._rbp]
+    mov esi, [eax + vcpu_state._rsi]
+    mov edi, [eax + vcpu_state._rdi]
+    mov eax, [eax + vcpu_state._rax]
+    je .resume
+    vmlaunch
+    jmp .exit_entry_fail
+.resume:
+    vmresume
+    jmp .exit_entry_fail
+.exit_entry:
+    push edi
+    mov edi, [esp+4]
+    mov [edi + vcpu_state._rax], eax
+    mov [edi + vcpu_state._rcx], ecx
+    mov [edi + vcpu_state._rdx], edx
+    pop ecx
+    mov [edi + vcpu_state._rbx], ebx
+    mov [edi + vcpu_state._rbp], ebp
+    mov [edi + vcpu_state._rsi], esi
+    mov [edi + vcpu_state._rdi], ecx
+.exit_entry_fail:
+    ; pop the state
+    pop eax
+    pop ebx
+    pop eax
+    pop ebp
+    pop edi
+    pop esi
+    pop edx
+    pop ecx
+    vmx_check
+    popfd
+    ret
+%else
+    %error "Unimplemented function"
+%endif
+
+function vmx_get_rip, 0
+    mov reg_ret, asm_vmxrun.exit_entry
```

Comment:
=> `lea reg_ret, [rel asm_vmxrun.exit_entry]`

This is enough to fix the macOS kernel panic! (A lot of credit goes to @junxiaoc )

Basically, the use of `lea` with the [REL keyword](https://www.nasm.us/doc/nasmdoc6.html#section-6.2.1) causes the address to be computed using RIP-relative addressing. The disassembly looks like this:

```asm
0000000000004a62        48 8d 05 7a ff ff ff    leaq    asm_vmxrun.exit_entry(%rip), %rax
```

Although `rel` is specific to 64-bit mode, it seems to be just ignored for the Win32 build, and this `lea` approach works across all platforms we support. Nevertheless, for readability (especially since `mov` is more intuitive than `lea`), we could save the `mov` approach for 32-bit.

Reply:
Please feel free to incorporate
[haxm-PR64-mac-kernel-panic-fix.zip](https://github.com/intel/haxm/files/2201035/haxm-PR64-mac-kernel-panic-fix.zip) into your 3rd commit.

Reply:
@raphaelning @junxiaoc Thank you so much for debugging+fixing the issue, and also for your patience with this PR. :-)

I'll add the patch and squash everything into the 3rd commit (it will be big, but as mentioned earlier it's convenient to avoid breaking git-bisect).